qe/drivers.py

```diff
@@ -1707,3 +1732,683 @@ def selection_step(
+
+    def explicit_matrix_product(self, V_k):
+        """Function to explicit compute matrix-matrix product W_ik = H_i * V_k.
+
+        :param H_i: local (self.local_size \times n) portion of the Hamiltonian, cached.
+        :param V_k: full column basis for trial subspace (n \times k), as a numpy array
+
+        :return W_ik: locally computed chunk of matrix-matrix product (self.local_size \times k), as a numpy array
+        """
+
+        return self.H_i_one_electron.explicit_1e_matrix_product(
+            V_k
+        ) + self.H_i_two_electrons.explicit_2e_matrix_product(V_k)
+
+
+@dataclass
+class Davidson_manager(object):
+    """A matrix-free implementation of Davidson's method in parallel.
+    All matrix products involving the Hamiltonian are computed implicitly and on-the-fly.
+
+    References:
+    * `A Parallel Davidson-Type Algorithm for Several Eigenvalues' [L. Borges, S. Oliveira, 1998]
+    * `The Davidson Method' [M. Crouzeix, B. Philippe, M. Sadkane, 1994]
+
+    Each process will have local access to d_one_e_integral, d_two_e_integral, E0,
+    and psi_full to build H_i on the fly.
+    :param d_two_e_integral: Dictionary of two-electron integrals
+    :param psi_full: Current trial wave-function, list of determinants. Problem size is n = len(psi_full)
+    """
+
+    #    d_one_e_integral: One_electron_integral
+    #    d_two_e_integral: Two_electron_integral
+    #    psi_full: Psi_det
+    #    E0: Energy
+
+    def __init__(
+        self,
+        comm,
+        #        psi_full: Psi_det,
+        #        d_one_e_integral: One_electron_integral,
+        #        d_two_e_integral: Two_electron_integral,
+        #        E0: Energy,
+        problem_size,
+    ):
+
+        self.comm = comm
+        self.world_size = self.comm.Get_size()  # No. of processes running
+        self.rank = self.comm.Get_rank()  # Rank of current process
+        self.MPI_master_rank = 0  # Master rank
+        # Create instance of Hamiltonian_Manger class(). Dispatches wavefunction to worker processes
+        #        self.Hamiltonian_manager = Hamiltonian_manager(
+        #            psi_full, d_one_e_integral, d_two_e_integral, E0, comm
+        #        )
+        # For easier reference, full problem size
+        self.full_size = problem_size
+        # Set local problem size
+        floor = self.full_size // self.world_size
+        ceiling = floor + 1
+        remainder = self.full_size % self.world_size
+        self.distribution = [ceiling] * remainder + [floor] * (self.world_size - remainder)
+        self.local_size = self.distribution[self.rank]  # Size of local chunk of Hamiltonian
+        if self.rank == 0:
+            print(
+                f"Distribution of work: {self.distribution}, Local problem size: {self.local_size}"
+            )
+        # Compute offsets (start of the local section) for all nodes
+        self.offsets = [0] + list(accumulate(self.distribution))
+        del self.offsets[-1]
```

Comment:
```suggestion
        self.offsets = [0] + list(accumulate(self.distribution[:-1]))
```
maybe a  little cleaner